### PR TITLE
Modified expand_path in config_parser to add environment var for mypy_path

### LINKED ISSUE #7967: Treat relative paths in config file relative to config file, not cwd
When using `mypy_path = src` it should be treated relative to the config file (if any), not the current working directory.
This also applies to the default of `cache_dir` (`.cache_dir`), which currently gets created in the current working directory always.

I.e. a detected/found (project) config file could be treated as the root for options with relative defaults also.

As for `mypy_path` this could be achieved in the type callback, which then however would need the config file / root as a new argument:
https://github.com/python/mypy/blob/84126ab9985844d8cda3809688034af895fdaf7c/mypy/config_parser.py#L75

As for global user config files, it should treat relative files therein relative to it, but it should not be used as a root for default settings like cache_dir.

### PR DESCRIPTION
**Pull request by ashutosh1919**
Possibly fixes #7967 and #7968 .
@JukkaL, @ethanhs and @blueyed , Please review this PR.

### CODE REVIEW

File: mypy/config_parser.py [hunk 1]
```diff
@@ -39,7 +39,12 @@ def expand_path(path: str) -> str:
     the provided path.
     """
 
-    return os.path.expandvars(os.path.expanduser(path))
+    expanded_path = os.path.expandvars(os.path.expanduser(path))
+    if 'MYPY_CONFIG_FILE_DIR' not in os.environ:
+        os.environ['MYPY_CONFIG_FILE_DIR'] = expanded_path
+    else:
+        os.environ['MYPY_CONFIG_FILE_DIR'] += (os.pathsep + expanded_path)
+    return expanded_path
```
Comment (blueyed) [inline mypy/config_parser.py]: That looks wrong to me: `expand_path` appears to be a generic method.
It should probably be done here instead: https://github.com/python/mypy/pull/8574/files#diff-09533a8b2bebadfae430681eb15b01c9R120 - i.e. setting it to the current file before parsing it.

Comment (ashutosh1919) [reply mypy/config_parser.py]: The link you gave above points to changed code itself. I don't understand your comment. Please, can you elaborate about where should the code be changed ?

Comment (blueyed) [reply mypy/config_parser.py]: @ashutosh1919 
Needs to be expanded via clicking, sorry.
https://github.com/python/mypy/blob/00e8024ffd187bdbddeb71fcced53cd2912a9410/mypy/config_parser.py#L120
But that is just a guess.
The idea would be to define it before parsing a config file, then it gets / can be expanded.

Comment (ashutosh1919) [reply mypy/config_parser.py]: @blueyed , Please take a look at it now.

**Comment by ashutosh1919** [comment]
@blueyed , I am not able to re-request review here. Please review.

File: mypy/config_parser.py [hunk 2]
```diff
@@ -112,6 +112,10 @@ def parse_config_file(options: Options, set_strict_flags: Callable[[], None],
         if not os.path.exists(config_file):
             continue
         try:
+            if 'MYPY_CONFIG_FILE_DIR' not in os.environ:
+                os.environ['MYPY_CONFIG_FILE_DIR'] = config_file
```
Comment (ilevkivskyi) [inline mypy/config_parser.py]: Why are you setting the _file_ name as a directory? Shouldn't you use something like `os.path.dirname()`?

Comment (ashutosh1919) [reply mypy/config_parser.py]: Yes you are right. Thanks for noticing. I will change it in next commit.

Comment (ashutosh1919) [reply mypy/config_parser.py]: done

**Comment by ashutosh1919** [comment]
@ilevkivskyi , please review.

**Comment by ilevkivskyi** [comment]
TBH, I don't totally understand the logic/reasoning behind this so I will leave it to someone else to review.

**Comment by ashutosh1919** [comment]
cc @JukkaL

**Comment by hauntsaninja** [comment]
Closing, since we merged #9414 to address the same issues

**Pull request closed without merging**